src/Microsoft.TestPlatform.Common/Filtering/FilterHelpers.cs

```diff
@@ -0,0 +1,231 @@
+            {
+                var current = str[i];
+
+                if (last == EscapePrefix)
+                {
+                    // Don't check if `current` is one of the special characters here.
+                    // Instead, we blindly let any character follows '\' pass though and 
+                    // relies on `FiltetrHelpers.Unescape` to report such errors.
+                    tokenBuilder.Append(current);
+
+                    if (current == EscapePrefix)
+                    {
+                        // We just encountered "\\" (escaped '\'), this will set last to '\0' 
+                        // so the next char will not be treated as a suffix of escape sequence.
+                        current = '\0';
+                    }
+                }
+                else
+                {
+                    switch (current)
+                    {
+                        case '(':
+                        case ')':
+                        case '&':
+                        case '|':
+                            if (tokenBuilder.Length > 0)
+                            {
+                                yield return tokenBuilder.ToString();
+                                tokenBuilder.Clear();
+                            }
+                            yield return current.ToString();   
+                            break;
+
+                        default:
+                            tokenBuilder.Append(current);
+                            break;
+                    }      
+                }
+
+                last = current;
+            }
+
+            if (tokenBuilder.Length > 0)
+            {
+                yield return tokenBuilder.ToString();
+            }
+        }
+
+        internal static IEnumerable<string> TokenizeFilterConditionString(string str)
+        {
+            if (str == null)
+            {
+                throw new ArgumentNullException(nameof(str));
+            }
+
+            StringBuilder tokenBuilder = new StringBuilder();
+
+            var last = '\0';
+            for (int i = 0; i < str.Length; ++i)
+            {
+                var current = str[i];
+
+                if (last == FilterHelpers.EscapePrefix)
+                {
+                    // Don't check if `current` is one of the special characters here.
+                    // Instead, we blindly let any character follows '\' pass though and 
+                    // relies on `FiltetrHelpers.Unescape` to report such errors.
+                    tokenBuilder.Append(current);
+
+                    if (current == FilterHelpers.EscapePrefix)
+                    {
+                        // We just encountered "\\" (escaped '\'), this will set last to '\0' 
+                        // so the next char will not be treated as a suffix of escape sequence.
+                        current = '\0';
+                    }
+                }
+                else
+                {
+                    switch (current)
```

Comment:
Just put a comment telling the allowed special tokens, eg "=", "!=", "!", "~"

Reply:
When do we expect just "!" ? Can you share an example, also add to tests.

Reply:
We don't expect '!', but it's not handled in tokenizer. Instead, an exception is thrown in `Condition.GetOperator`.